graylog2-web-interface/src/views/types.d.ts

```diff
@@ -0,0 +1,118 @@
+import { WidgetActionType } from 'views/components/widgets/Types';
+import { Creator } from 'views/components/sidebar/create/AddWidgetButton';
+import { ViewHook } from 'views/logic/hooks/ViewHook';
+import WidgetConfig from 'views/logic/widgets/WidgetConfig';
+import FieldTypeMapping from 'views/logic/fieldtypes/FieldTypeMapping';
+import { Completer } from 'views/components/searchbar/SearchBarAutocompletions';
+import { Result } from 'views/components/widgets/Widget';
+import { Widgets } from 'views/stores/WidgetStore';
+
+interface EditWidgetComponentProps<Config extends WidgetConfig = WidgetConfig> {
+  children: React.ReactNode,
+  config: Config,
+  editing: boolean;
+  id: string;
+  type: string;
+  fields: Immutable.List<FieldTypeMapping>,
+  onChange: (newConfig: Config) => void,
+}
+
+interface WidgetComponentProps<Config extends WidgetConfig = WidgetConfig> {
+  config: Config;
+  data: { [key: string]: Result };
+  editing: boolean;
+  fields: Immutable.List<FieldTypeMapping>;
+  filter: string;
+  height: number;
+  width: number;
+  queryId: string;
+  onConfigChange: (newConfig: Config) => Promise<Widgets>;
+  setLoadingState: (loading: boolean) => void;
+  title: string;
+  toggleEdit: () => void;
+  type: string;
+  id: string;
+}
+
+export interface WidgetExport {
+  type: string;
+  displayName?: string;
+  defaultHeight?: number;
+  defaultWidth?: number;
+  visualizationComponent: React.ComponentType<WidgetComponentProps>;
+  editComponent: React.ComponentType<EditWidgetComponentProps>;
+  needsControlledHeight: (widget: Widget) => boolean;
+  searchResultTransformer?: (data: Array<unknown>, widget: Widget) => unknown;
+  searchTypes: (widget: Widget) => Array<any>;
+  titleGenerator?: (widget: Widget) => string;
+  reportStyle?: () => Record<string, any>;
+}
+
+interface VisualizationConfigProps {
+  config: WidgetConfig;
+  onChange: (newConfig: WidgetConfig) => void;
+}
+
+interface VisualizationConfigType {
+  type: string;
+  component: React.ComponentType<VisualizationConfigProps>;
+}
+
+interface VisualizationType {
+  type: string;
+  displayName: string;
+  component: VisualizationComponent;
+}
+
+interface ResultHandler<T, R> {
+  convert: (result: T) => R;
+}
+interface SearchType {
+  type: string;
+  handler: ResultHandler;
+  defaults: {};
+}
+
+declare module 'graylog-web-plugin/plugin' {
+  export interface PluginExports {
+    creators?: Array<Creator>;
+    enterpriseWidgets?: Array<WidgetExport>;
+    fieldActions?: Array<ActionDefinition>;
+    searchTypes?: Array<SearchType>;
+    valueActions?: Array<ActionDefinition>;
+    'views.completers'?: Array<Completer>;
+    'views.elements.header'?: Array<React.ComponentType>;
+    'views.elements.queryBar'?: Array<React.ComponentType>;
+    'views.hooks.executingView'?: Array<ViewHook>,
+    'views.hooks.loadingView'?: Array<ViewHook>,
+    'views.hooks.searchRefresh'?: Array<SearchRefreshCondition>;
+    'views.overrides.widgetEdit'?: Array<>;
```

Comment:
Could we use the same type definition like in `WidgetOverrideElements` for `views.overrides.widgetEdit`? `Array<React.ComponentType<OverrideProps>>`

Reply:
:white_check_mark: